File: implementations/python/ucs7.py
"""7bit-ucs encoder/decoder (single-file implementation).

Part of the 7bit-ucs project.
Author: 刘占江 (Liu Zhanjiang)
Email:  jny7@163.com
GitHub: https://github.com/uponcn/7bit-ucs

SPDX-License-Identifier: MIT
"""

from __future__ import annotations

from typing import Tuple

MAX_CODE_POINT = 0x10FFFF


class Ucs7Error(ValueError):
    """Raised on invalid 7bit-ucs input or code point."""


def encode(cp: int) -> bytes:
    """Encode a Unicode code point into 7bit-ucs bytes."""
    if not (0 <= cp <= MAX_CODE_POINT):
        raise Ucs7Error(f"invalid code point: U+{cp:04X}")
    if cp < 0x80:
        return bytes([cp])
    if cp < 0x4000:
        return bytes([0x80 | (cp >> 7), cp & 0x7F])
    return bytes(
        [
            0x80 | (cp >> 14),
            0x80 | ((cp >> 7) & 0x7F),
            cp & 0x7F,
        ]
    )


def decode(data: bytes, offset: int = 0) -> Tuple[int, int]:
    """Decode one 7bit-ucs character.

    Returns (code_point, bytes_consumed).
    Raises Ucs7Error on invalid input.
    """
    acc = 0
    for i in range(3):
        if offset + i >= len(data):
            raise Ucs7Error("truncated sequence")
        b = data[offset + i]
        acc = (acc << 7) | (b & 0x7F)
        if b < 0x80:
            if i == 1 and acc < 0x80:
                raise Ucs7Error("overlong encoding")
            if i == 2 and acc < 0x4000:
                raise Ucs7Error("overlong encoding")
            if acc > MAX_CODE_POINT:
                raise Ucs7Error("code point out of range")
            return acc, i + 1
    raise Ucs7Error("sequence longer than 3 bytes")


def encode_str(s: str) -> bytes:
    """Encode a Python string into 7bit-ucs bytes."""
    out = bytearray()
    for ch in s:
        out.extend(encode(ord(ch)))
    return bytes(out)


def decode_str(data: bytes) -> str:
    """Decode 7bit-ucs bytes into a Python string."""
    out = []
    pos = 0
    while pos < len(data):
        cp, n = decode(data, pos)
        out.append(chr(cp))
        pos += n
    return "".join(out)


def find_start(data: bytes, pos: int) -> int:
    """Return the index of the first byte of the character containing pos."""
    if pos == 0:
        return 0
    if data[pos] < 0x80:
        if data[pos - 1] < 0x80:
            return pos
        if pos < 2 or data[pos - 2] < 0x80:
            return pos - 1
        return pos - 2
    if data[pos - 1] < 0x80:
        return pos
    return pos - 1


__all__ = [
    "MAX_CODE_POINT",
    "Ucs7Error",
    "encode",
    "decode",
    "encode_str",
    "decode_str",
    "find_start",
]


if __name__ == "__main__":
    samples = [
        (0x0041, b"\x41"),
        (0x00E9, b"\x81\x69"),
        (0x0915, b"\x92\x15"),
        (0x4E2D, b"\x81\x9C\x2D"),
        (0x1F600, b"\x87\xEC\x00"),
    ]
    for cp, expected in samples:
        got = encode(cp)
        assert got == expected, (hex(cp), got, expected)
        back, n = decode(got)
        assert back == cp and n == len(got)

    s = "Hello \u4e16\u754c \U0001F600 \u0645\u0631\u062d\u0628\u0627"
    assert decode_str(encode_str(s)) == s

    # self-sync
    buf = encode_str("中😀क")
    for pos in range(len(buf)):
        start = find_start(buf, pos)
        _, used = decode(buf, start)
        assert start + used > pos

    print("all tests passed")

File: implementations/java/Ucs7.java
/*
 * Ucs7.java - 7bit-ucs encoder/decoder (single-file implementation)
 *
 * Part of the 7bit-ucs project.
 * Author: 刘占江 (Liu Zhanjiang)
 * Email:  jny7@163.com
 * GitHub: https://github.com/uponcn/7bit-ucs
 *
 * SPDX-License-Identifier: MIT
 *
 * Build and run:
 *     javac Ucs7.java && java Ucs7
 * or (JDK 11+):
 *     java Ucs7.java
 */

public final class Ucs7 {

    public static final int MAX_CODE_POINT = 0x10FFFF;

    private Ucs7() {
    }

    public static byte[] encode(int cp) {
        if (cp < 0 || cp > MAX_CODE_POINT) {
            throw new IllegalArgumentException(
                    String.format("invalid code point: U+%04X", cp));
        }
        if (cp < 0x80) {
            return new byte[] { (byte) cp };
        }
        if (cp < 0x4000) {
            return new byte[] {
                    (byte) (0x80 | (cp >> 7)),
                    (byte) (cp & 0x7F)
            };
        }
        return new byte[] {
                (byte) (0x80 | (cp >> 14)),
                (byte) (0x80 | ((cp >> 7) & 0x7F)),
                (byte) (cp & 0x7F)
        };
    }

    public static int[] decode(byte[] data, int offset) {
        int acc = 0;
        for (int i = 0; i < 3; i++) {
            if (offset + i >= data.length) {
                throw new IllegalArgumentException("truncated sequence");
            }
            int b = data[offset + i] & 0xFF;
            acc = (acc << 7) | (b & 0x7F);
            if (b < 0x80) {
                if (i == 1 && acc < 0x80) {
                    throw new IllegalArgumentException("overlong encoding");
                }
                if (i == 2 && acc < 0x4000) {
                    throw new IllegalArgumentException("overlong encoding");
                }
                if (acc > MAX_CODE_POINT) {
                    throw new IllegalArgumentException("code point out of range");
                }
                return new int[] { acc, i + 1 };
            }
        }
        throw new IllegalArgumentException("sequence longer than 3 bytes");
    }

    public static byte[] encodeString(String s) {
        byte[] buf = new byte[s.length() * 3];
        int pos = 0;
        for (int i = 0; i < s.length(); ) {
            int cp = s.codePointAt(i);
            i += Character.charCount(cp);
            byte[] enc = encode(cp);
            System.arraycopy(enc, 0, buf, pos, enc.length);
            pos += enc.length;
        }
        byte[] out = new byte[pos];
        System.arraycopy(buf, 0, out, 0, pos);
        return out;
    }

    public static String decodeString(byte[] data) {
        StringBuilder sb = new StringBuilder();
        int pos = 0;
        while (pos < data.length) {
            int[] r = decode(data, pos);
            sb.appendCodePoint(r[0]);
            pos += r[1];
        }
        return sb.toString();
    }

    public static int findStart(byte[] data, int pos) {
        if (pos == 0) {
            return 0;
        }
        int cur = data[pos] & 0xFF;
        int prev = data[pos - 1] & 0xFF;
        if (cur < 0x80) {
            if (prev < 0x80) {
                return pos;
            }
            if (pos < 2 || (data[pos - 2] & 0xFF) < 0x80) {
                return pos - 1;
            }
            return pos - 2;
        }
        if (prev < 0x80) {
            return pos;
        }
        return pos - 1;
    }

    public static void main(String[] args) {
        int[][] samples = {
                { 0x0041, 0x41 },
                { 0x00E9, 0x8169 },
                { 0x0915, 0x9215 },
                { 0x4E2D, 0x819C2D },
                { 0x1F600, 0x87EC00 }
        };
        for (int[] s : samples) {
            int cp = s[0];
            byte[] enc = encode(cp);
            int[] dec = decode(enc, 0);
            if (dec[0] != cp || dec[1] != enc.length) {
                throw new AssertionError("roundtrip failed for U+"
                        + Integer.toHexString(cp));
            }
        }

        // Overlong rejection
        try {
            decode(new byte[] { (byte) 0x80, 0x00 }, 0);
            throw new AssertionError("overlong not rejected");
        } catch (IllegalArgumentException ok) {
            // expected
        }

        // Too long rejection
        try {
            decode(new byte[] { (byte) 0x80, (byte) 0x80, (byte) 0x80, 0x00 }, 0);
            throw new AssertionError("too-long not rejected");
        } catch (IllegalArgumentException ok) {
            // expected
        }

        // Truncated rejection
        try {
            decode(new byte[] { (byte) 0x80 }, 0);
            throw new AssertionError("truncated not rejected");
        } catch (IllegalArgumentException ok) {
            // expected
        }

        // Self-sync
        byte[] buf = encodeString("\u4e2d\ud83d\ude00\u0915A");
        for (int pos = 0; pos < buf.length; pos++) {
            int start = findStart(buf, pos);
            int[] r = decode(buf, start);
            if (start + r[1] <= pos) {
                throw new AssertionError("self-sync failed at " + pos);
            }
        }

        // String roundtrip
        String s = "Hello \u4e16\u754c \ud83d\ude00";
        if (!s.equals(decodeString(encodeString(s)))) {
            throw new AssertionError("string roundtrip failed");
        }

        System.out.println("all tests passed");
    }
}
